### PR TITLE
Fixed #5908 -- Added {% resetcycle %} template tag

### PR DESCRIPTION
**Pull request by sergei-maertens**
Added the {% resetcycle %} template tag to restart the
{% cycle %} tag from its first value when it's next encountered.

Thanks to Simon Litchfield for the report, Uninen for the initial
patch, akaihola, jamesp, b.schube and Florian Appoloner for
subsequent patches, tests and documentation.

### CODE REVIEW

File: django/template/defaulttags.py [hunk 1]
```diff
@@ -1220,6 +1239,34 @@ def regroup(parser, token):
...
+
+    if len(args) > 2:
+        raise TemplateSyntaxError("%r tag accepts at most one argument." % args[0])
+
+    if len(args) == 2:
+        name = args[1]
+        try:
+            return ResetCycleNode(parser._named_cycle_nodes[name])
+        except AttributeError:
+            raise TemplateSyntaxError("No named cycles in template. %r is not defined" % name)
```
Comment (profuel) [inline django/template/defaulttags.py]: I would suggest replace for second part with same from `KeyError`:
`Named cycle %r does not exist`
It's more transparent then fast that 'ticks is not defined', for cycle called ticks.

Comment (sergei-maertens) [inline django/template/defaulttags.py]: I *think* that the intended important bit of the message here is that there are no named cycles at all in the template, and thus the named cycle is not defined. I'm happy to change it, as it is a bit more readable, as long as it doesn't distract fromt the other error messages where there are named cycles available, but not this one.

@timgraham your thoughts?

File: docs/ref/templates/builtins.txt [hunk 2]
```diff
@@ -185,6 +185,9 @@ call to ``{% cycle %}`` doesn't specify ``silent``::
     {% cycle 'row1' 'row2' as rowcolors silent %}
     {% cycle rowcolors %}
 
+You can use the `resetcycle`_ tag to reset a ``{% cycle %}`` tag to restart
```
Comment (timgraham) [inline docs/ref/templates/builtins.txt]: use :ttag: for linking

File: docs/ref/templates/builtins.txt [hunk 3]
```diff
@@ -971,6 +974,55 @@ attribute, allowing  you to group on the display string rather than the
 ``{{ country.grouper }}`` will now display the value fields from the
 ``choices`` set rather than the keys.
 
+.. templatetag:: resetcycle
+
+``resetcycle``
+--------------
+
```
Comment (timgraham) [inline docs/ref/templates/builtins.txt]: `.. versionadded:: 1.11`

File: django/template/defaulttags.py (hunk 1 above)
Comment (timgraham) [inline django/template/defaulttags.py]: It doesn't seem particularly necessary to mention "No named cycles in template." .I think "Named cycle %r does not exist." seems like enough.

File: django/template/defaulttags.py [hunk 4]
```diff
@@ -586,6 +601,9 @@ def cycle(parser, token):
     # that names are only unique within each template (as opposed to using
     # a global variable, which would make cycle names have to be unique across
     # *all* templates.
+    #
+    # It also keeps the last node in the parser, to be able to reset it
```
Comment (timgraham) [inline django/template/defaulttags.py]: "reset it with {% resetcycle %}"? (and add period)

File: django/template/defaulttags.py [hunk 5]
```diff
@@ -595,11 +613,11 @@ def cycle(parser, token):
     if len(args) == 2:
         # {% cycle foo %} case.
         name = args[1]
-        if not hasattr(parser, '_namedCycleNodes'):
+        if not hasattr(parser, '_named_cycle_nodes'):
```
Comment (timgraham) [inline django/template/defaulttags.py]: I'd do this renaming in a separate commit to keep the diff for the feature cleaner.

File: django/template/defaulttags.py [hunk 6]
```diff
@@ -87,6 +87,12 @@ def render(self, context):
             return ''
         return render_value_in_context(value, context)
 
+    def reset(self, context):
+        """
+        Resets the cycle iteration back to the beginning.
```
Comment (timgraham) [inline django/template/defaulttags.py]: use pep257 docstring verb style "Reset... "

File: tests/template_tests/syntax_tests/test_resetcycle.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+from django.test import SimpleTestCase
+
+from ..utils import setup
+
+
+class ResetCycleTagTests(SimpleTestCase):
+
+    @setup({'resetcycle01': "{% resetcycle %}"})
+    def test_resetcycle01(self):
+        with self.assertRaises(TemplateSyntaxError):
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/test_resetcycle.py]: please use `assertRaisesMessage`

File: docs/releases/1.11.txt [hunk 8]
```diff
@@ -230,6 +230,9 @@ Templates
   supports context processors by setting the ``'context_processors'`` option in
   :setting:`OPTIONS <TEMPLATES-OPTIONS>`.
 
+* Added a :ttag:`resetcycle` template tag which allows the user to reset the
```
Comment (timgraham) [inline docs/releases/1.11.txt]: to allow resetting the

**Comment by sergei-maertens** [comment]
@timgraham I've processed the remarks and made the necessary changes

**Comment by timgraham** [comment]
merged in 32c02f2a0eaf66e4744f9904eb2a743b87300257, thanks!
